lib/edition_assertions.rb

```diff
@@ -14,11 +14,23 @@ def initialize(edition, assertion)
     end
   end
 
+  class FeatureError < Error
+    def initialize(edition, assertion)
+      super(edition, "edition.document_type #{assertion}")
+    end
+  end
+
   def assert_edition_state(edition, options = {}, &block)
     return if block.call(edition)
 
     assertion = options[:assertion]
     assertion ||= block.to_s if block.to_s =~ /&:/ # &:editable?
     raise StateError.new(edition, assertion)
   end
+
+  def assert_edition_feature(edition, &block)
+    return if block.call(edition.document_type)
+
+    raise FeatureError.new(edition, block.to_s)
```

Comment:
I'm wondering here if we're only expecting this to be called with `&:lead_image` style blocks? I'm basing that on us not doing the `if block.to_s =~ /&:/` check and so we'd end up with "#<Proc:0x00007fd" style output (if I understand right) for a customised block.

So this makes me think maybe we actually want to just change &block to be a symbol of the feature name and not worry about blocks and instead do `edition.document_type.public_send(feature)`.

Reply:
This is actually the approach I was assuming and started with. I wrote in the commit how this was awkward, because I wanted to make the assertion method robust for different arguments; at the moment, we only support `:lead_image`. It seemed simpler and more consistent to directly tie the use of the assertion to the invocation of a method.

Alternatives I thought of:

   - Have a `ASSERTABLE_FEATURES` constant and raise if outside of this
   - Check the class of the 'flag' and raise if it's not a boolean

I think this is a symptom of postponing https://github.com/alphagov/content-publisher/issues/1676, which is fine, as all the alternatives are relatively simple.

Do you have any preferences or other suggestions?

Reply:
Updated as discussed.